### PR TITLE
partial create issue resolved for volume, storage pool and snapshot

### PR DESCRIPTION
**Pull request by AnikaAgiwal2711**
# Description
A few sentences describing the overall goals of the pull request's commits.

# GitHub Issues
List the GitHub issues impacted by this PR:

| GitHub Issue # |
| -------------- |
| |

# Checklist:

- [ ] I have performed a self-review of my own code to ensure there are no formatting, vetting, linting, or security issues
- [ ] I have verified that new and existing unit tests pass locally with my changes
- [ ] I have not allowed coverage numbers to degenerate
- [ ] I have maintained at least 80% code coverage
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] Backward compatibility is not broken

# How Has This Been Tested?
Please describe the tests that you ran to verify your changes. Please also list any relevant details for your test configuration

- [ ] Test A
- [ ] Test B

### CODE REVIEW

**Comment by AnikaAgiwal2711** [comment]
run e2e test

**Comment by AnikaAgiwal2711** [comment]
run e2e test

**Comment by AnikaAgiwal2711** [comment]
run e2e test

File: powerflex/provider/snapshot_resource.go [hunk 1]
```diff
@@ -203,6 +203,9 @@ func (r *snapshotResource) Create(ctx context.Context, req resource.CreateReques
 		errMsg["desired_retention/retention_unit"] = "The specified snapshot can't be retained due to failure in creation."
 	}
 	if (len(errMsg) == 0) && !plan.DesiredRetention.IsNull() {
+		if plan.DesiredRetention.ValueInt64() <= 0 {
+			errMsg["desired_retention/retention_unit"] = "Value of desired retention can't be negative."
```
Comment (doriac11) [inline powerflex/provider/snapshot_resource.go:207]: Why not return here if an error is hit?

File: powerflex/provider/snapshot_resource.go [hunk 2]
```diff
@@ -217,10 +220,13 @@ func (r *snapshotResource) Create(ctx context.Context, req resource.CreateReques
 		return
 	}
 	snap = snapResponse[0]
-	dgs := helper.RefreshState(snap, &plan)
-	resp.Diagnostics.Append(dgs...)
-	diags = resp.State.Set(ctx, &plan)
-	resp.Diagnostics.Append(diags...)
+	if len(errMsg) == 0 {
```
Comment (shenda1) [inline powerflex/provider/snapshot_resource.go:223]: The subsequent call after creating snapshot is modifying the size. If this fails, still results in partial creation.

File: powerflex/provider/storagepool_resource.go [hunk 3]
```diff
@@ -415,6 +415,18 @@ func (r *storagepoolResource) Create(ctx context.Context, req resource.CreateReq
 		}
 	}
 
+	if resp.Diagnostics.HasError() {
+		err = pd.DeleteStoragePool(plan.Name.ValueString())
```
Comment (shenda1) [inline powerflex/provider/storagepool_resource.go:419]: There are multiple operations after creating storage pool. If any one operation is failed, storage pool will be deleted then  why should we continue with operations?

Comment (AnikaAgiwal2711) [reply powerflex/provider/storagepool_resource.go:414]: Done

File: powerflex/provider/snapshot_resource.go (hunk 2 above)
Comment (AnikaAgiwal2711) [reply powerflex/provider/snapshot_resource.go]: Done

File: powerflex/provider/snapshot_resource.go (hunk 1 above)
Comment (AnikaAgiwal2711) [reply powerflex/provider/snapshot_resource.go:203]: Done

**Comment by AnikaAgiwal2711** [comment]
image
image
image

**Review by shenda1 — APPROVED** [review]

**Comment by AnikaAgiwal2711** [comment]
run e2e test VolumeResource,StoragepoolResource,SnapshotResource

**Comment by shenda1** [comment]
run e2e test VolumeResource,StoragepoolResource,SnapshotResource

**Review by Krunal-Thakkar — APPROVED** [review]

**Pull request merged**